[BEAM-2732][BEAM-4028] Logging relies on StateSampler for context

pabloem opened the pull request:
The Logging module will no longer implement its own context tracking for step and stage names.

Because it no longer plays part in high-performance operations, I'm removing it and its cython annotations.

Passing PostCommit: https://builds.apache.org/job/beam_PostCommit_Python_Verify/4985/

pabloem commented:
Run Python ValidatesRunner

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
r: @aaltay 
This unifies context management in Python, which simplifies further feature work, and also expands the use of NameContext, which should improve the separation of runner and sdk harness.

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
r: @charlesccychen 
This unifies context management in Python, so that logging will use the state sampler to retrieve its current state. Also, NameContext is used more widely to improve step name management.

charlesccychen commented on sdks/python/apache_beam/runners/common.py:
```diff
@@ -539,19 +529,14 @@ def __init__(self,
       tagged_receivers: a dict of tag name to Receiver objects
       step_name: the name of this step
-      logging_context: a LoggingContext object
+      logging_context: DEPRECATED
```
Can you add a JIRA to remove this?

pabloem replied:
Added BEAM-4728.

charlesccychen commented on sdks/python/apache_beam/runners/worker/bundle_processor.py:
```diff
@@ -59,13 +59,12 @@
+        name_context, None, counter_factory, state_sampler)
     self.windowed_coder = windowed_coder
     self.windowed_coder_impl = windowed_coder.get_impl()
-    self.step_name = step_name
```
Is this deletion intentional?  If so, can you add a comment / JIRA reference to clean up step_name in the arguments?

pabloem replied:
This is part of my goal with BEAM-4028. The step name is meant to only be retrievable through the name context.

charlesccychen commented on sdks/python/apache_beam/runners/worker/logger.py:
```diff
@@ -49,7 +48,7 @@ def get_data(self):
 
 
-class PerThreadLoggingContext(LoggingContext):
+class PerThreadLoggingContext(object):
```
Are we able to get rid of this class entirely?  It looks like you removed the only usage in operations.py.

charlesccychen commented on sdks/python/apache_beam/runners/worker/logger.py:
```diff
@@ -34,7 +34,6 @@ class _PerThreadWorkerData(threading.local):
 
   def __init__(self):
     super(_PerThreadWorkerData, self).__init__()
-    # TODO(robertwb): Consider starting with an initial (ignored) ~20 elements
```
Accidental deletion?

pabloem replied:
The logging context will be removed once it's no longer useful (after it's removed from Google code) so optimizations should not be considered anymore. I'll remove it ASAP as part of BEAM-4728

charlesccychen commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -62,12 +64,22 @@ def __init__(self, prefix, counter_factory,
     self.states_by_name = {}
     self._prefix = prefix
-    self._counter_factory = counter_factory
+    self._counter_factory = counter_factory or CounterFactory()
```
It looks like the second branch is only used by tests.  Can we have the tests pass empty `CounterFactory()`s instead of adding this optional behavior in the actual code?

pabloem commented on sdks/python/apache_beam/runners/worker/operation_specs.py:
```diff
@@ -376,9 +376,9 @@ def __init__(self, operations, stage_name,
+
     self.operations = operations
     self.stage_name = stage_name
-    # TODO(BEAM-4028): Remove arguments other than name_contexts.
```
Oops. Not obsolete. Good catch!

pabloem commented:
Also removed the extra path in `statesampler` for `CounterFactory`

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
Results of the `map_fn_microbenchmark`, for some floating point reason, it gives per-element cost as zero, but if you observe row-per-row, the cost with changes is slightly lower than the cost on master.

On master:
```
     1 element  0.798268 sec
  1001 elements 1.01165 sec
  2001 elements 1.05419 sec
  3001 elements 1.1398 sec
  4001 elements 1.37623 sec
  5001 elements 1.47872 sec
  6001 elements 1.68769 sec
  7001 elements 1.68809 sec
  8001 elements 1.8503 sec
  9001 elements 2.0606 sec
Fixed cost   0.8104043092640963
Per-element  0.0
R^2          0.9845043457059202
```

With these changes:
```
     1 element  0.796835 sec
  1001 elements 0.952501 sec
  2001 elements 1.01314 sec
  3001 elements 1.17117 sec
  4001 elements 1.31416 sec
  5001 elements 1.3791 sec
  6001 elements 1.54986 sec
  7001 elements 1.68663 sec
  8001 elements 1.71841 sec
  9001 elements 1.93632 sec
Fixed cost   0.8011857992764676
Per-element  0.0
R^2          0.9922532858766798

```

charlesccychen commented:
Thanks Pablo!  This LGTM.  Can you rebase?  It looks like this is a great performance improvement too, cutting down processing overhead.

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
Tests passing. I'll squash and merge this after lunch.

pabloem commented:
Squashed commits and resolved conflicts. Reruning tests.

pabloem commented:
Run Python Dataflow ValidatesRunner

pabloem commented:
Merged. Thanks @charlesccychen for reviewing the large-ish change : )
